### PR TITLE
DRLG Fixes, seems to be playable everywhere now !

### PR DESCRIPTION
**Pull request by Lectem**
I went through all the Normal mode levels (in a single save though) to test, and didn't see any obvious differences nor crashes with the fixes in this PR.

There could very well be a few bugs remaining, but this is way better than what we had before.
I realize it is (again) a big PR, so I'll just merge it after some time if nobody could review it.

### CODE REVIEW

File: source/D2Common/include/Drlg/D2DrlgRoomTile.h [hunk 1]
```diff
@@ -5,9 +5,20 @@
...
+	PACKEDTILEINFO_FLOOR = 0b10,
+	PACKEDTILEINFO_STRUCTUREPARTMASK = PACKEDTILEINFO_ROOF | PACKEDTILEINFO_WALL | PACKEDTILEINFO_FLOOR,
+	PACKEDTILEINFO_ANIMATEDTILE = 1 << 27,
+};
 
 #pragma pack()
 
+// Helper functions
+static uint32_t GetTileStyleFromPackedTileInformation(uint32_t nTileInformation) { return (nTileInformation >> 20) & 0x3F; }
+static uint32_t GetTileSequenceFromPackedTileInformation(uint32_t nTileInformation) { return BYTE1(nTileInformation); }
```
Comment (Mw95) [inline source/D2Common/include/Drlg/D2DrlgRoomTile.h:21]: Theoretically, these could be used at way more places than they are currently

**Review by Mw95 — COMMENTED** [review]
There seems to be a minor bug in DrlgDrlgAnim.cpp, everything else looks fine to me.

In general, I'd prefer smaller commits to simplify the review. In my opinion, non-critical changes like renaming or reformatting something should be done in a separate commit which then doesn't necessarily need to be reviewed/put into a PR, though.

File: source/D2Common/include/Drlg/D2DrlgRoomTile.h (hunk 1 above)
Comment (Lectem) [reply source/D2Common/include/Drlg/D2DrlgRoomTile.h:21]: I think so too.

Mentor's note have the following table, which I need to confirm but seems to match.


<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Lectem/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Lectem/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink="#954F72">



flag | bit | enum eDS1Flags | comment | tiledata flag
-- | -- | -- | -- | --
0x00000001 | 0 | TILEFLAG_WALL | default   valid wall |  
0x00000002 | 1 | TILEFLAG_FLOOR | default   valid floor | 0x2000
0x00000004 | 2 | TILEFLAG_LOS | code-generated;   added on edges |  
0x00000008 | 3 | TILEFLAG_ENCLOSE | seems   to delimit an enclosure inside another area; | 0x04
0x00000010 | 4 | TILEFLAG_EXIT | arch   or doorway or gateway in a wall |  
0x00000020 | 5 | TILEFLAG_20 |   |  
0x00000040 | 6 | TILEFLAG_LAYER_BELOW | only   floor |  
0x00000080 | 7 | TILEFLAG_LAYER_ABOVE | wall   & floor | 0x01
0x00000100 | 8 | TILEFLAG_TILESUBINDEX |   |  
0x00000200 | 9 | TILEFLAG_TILESUBINDEX |   |  
0x00000400 | 10 | TILEFLAG_TILESUBINDEX |   |  
0x00000800 | 11 | TILEFLAG_TILESUBINDEX |   |  
0x00001000 | 12 | TILEFLAG_TILESUBINDEX |   |  
0x00002000 | 13 | TILEFLAG_TILESUBINDEX |   |  
0x00004000 | 14 | TILEFLAG_TILESUBINDEX | seems   to be subindex, too. AND 0x3F0FF00 is used to check if index 0 subindex 0 |  
0x00008000 | 15 | TILEFLAG_TILESUBINDEX | seems   to be subindex, too. AND 0x3F0FF00 is used to check if index 0 subindex 0 |  
0x00010000 | 16 | TILEFLAG_FILL_LOS | all   subtiles will get wall collision; | 0x80
0x00020000 | 17 | TILEFLAG_UNWALKABLE |   | 0x40
0x00040000 | 18 | TILEFLAG_OVERLAPPED_LAYER2 | code-generated;   set when at least Floor2 or Wall2  //   (nLayer - 1) << 0x12 |  
0x00080000 | 19 | TILEFLAG_OVERLAPPED_LAYER3 | code-generated;   set when has more walls, incl' of other orientations than usual; obj, shd,   tree, roof, lower |  
0x000C0000 | 19 | TILEFLAG_OVERLAPPED_LAYER4 | code-generated;   set when has more walls, incl' of other orientations than usual; obj, shd,   tree, roof, lower |  
0x00100000 | 20 | TILEFLAG_TILEINDEX | SHR   R32, 14  //  AND R32, 3F |  
0x00200000 | 21 | TILEFLAG_TILEINDEX |   |  
0x00400000 | 22 | TILEFLAG_TILEINDEX |   |  
0x00800000 | 23 | TILEFLAG_TILEINDEX |   |  
0x01000000 | 24 | TILEFLAG_TILEINDEX |   |  
0x02000000 | 25 | TILEFLAG_TILEINDEX |   |  
0x04000000 | 26 | TILEFLAG_REVEAL_HIDDEN | looks   like an upper wall brought to a layer in front | 0x20C
0x08000000 | 27 | TILEFLAG_SHADOW | this   layer is a shadow layer |  
0x10000000 | 28 | TILEFLAG_LINKAGE | near   wp, lvl links, paths // will never get hidden | 0x102
0x20000000 | 29 | TILEFLAG_OBJECT_WALL | wall   tiles with props; may be block reverb / other sounds (crates, barrels, tables   etc.) | 0x800
0x40000000 | 30 | TILEFLAG_40000000 |   |  
0x80000000 | 31 | TILEFLAG_HIDDEN |   | 0x08



</body>

</html>

File: source/D2Common/src/Drlg/DrlgDrlgAnim.cpp [hunk 2]
```diff
@@ -100,32 +94,36 @@ void __fastcall DRLGANIM_TestLoadAnimatedRoomTiles(D2RoomExStrc* pRoomEx, D2Drlg
...
-				pTileData = j->ppMapTileData[j->nCurrentFrame >> 8];
+				D2DrlgTileDataStrc* pTileData = j->ppMapTileData[AnimationFrameFixedPointToInteger(j->nCurrentFrame)];
 				if (pTileData)
 				{
-					pTileData->dwFlags |= 8;
+					pTileData->dwFlags |= MAPTILE_HIDDEN;
 				}
 
-				j->nCurrentFrame = (j->nCurrentFrame + j->nAnimationSpeed) % (j->nFrames << 8);
+				j->nCurrentFrame = (j->nCurrentFrame + j->nAnimationSpeed) % AnimationFrameFixedPointToInteger(j->nFrames);
```
[the remark this answers is not in the document]
Comment (Lectem) [reply source/D2Common/src/Drlg/DrlgDrlgAnim.cpp]: That was indeed a mistake, fixed.

File: source/D2Common/include/Drlg/D2DrlgRoomTile.h (hunk 1 above)
Comment (Lectem) [reply source/D2Common/include/Drlg/D2DrlgRoomTile.h:21]: I will investigate this in a future PR

**Pull request merged**